File: binary_tree_from_preorder_and_inorder.py
"""
Problem: Construct Binary Tree from Preorder and Inorder Traversal (No duplicates)
Leetcode: https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
Solution:
    - Use preorder to find root element at first index
    - Use inorder to find left and right elements of root index
Time Complexity: O(N), being used to traverse the preorder list
Space Complexity: O(N), being used to store the inorder list values and index's
"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    # class variables
    pre_idx = 0
    preorder = []
    inorder = []
    # hashmap to store inorder value and index's
    index_map = {}

    def buildTree(self, preorder, inorder) -> TreeNode:
        self.preorder = preorder
        self.inorder = inorder

        # Add inorder value and index's to index_map
        idx = 0
        for i in range(len(inorder)):
            self.index_map[inorder[i]] = i

        return self.helper(0, len(self.inorder) - 1)

    def helper(self, left, right):
        # Base case
        if left > right:
            return None

        # find the root from preorder list
        root_val = self.preorder[self.pre_idx]
        root = TreeNode(root_val)
        self.pre_idx += 1

        index = self.index_map[root_val]

        # create left subtree
        root.left = self.helper(left, index - 1)

        # create right subtree
        root.right = self.helper(index + 1, right)

        return root

File: binaryTreeFromPreorderAndInorder.java
/**
Problem: https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

/**
Approach 1: through array copy and searching for root in inorder array
TC: O(n^2)
SC: O(n^2)
*/
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder == null || inorder == null || preorder.length == 0 || inorder.length == 0) {
            return null;
        }
        
        TreeNode root = new TreeNode(preorder[0]);
        int rootIdx = -1;
        for (int i = 0; i < inorder.length; ++i) {
            if (inorder[i] == root.val) {
                rootIdx = i;
                break;
            }
        }
        
        
        int[] preorderLeft = Arrays.copyOfRange(preorder, 1, rootIdx + 1); // Arrays.copyOfRange ignores last element
        int[] preorderRight = Arrays.copyOfRange(preorder, rootIdx + 1, preorder.length);
        int[] inorderLeft = Arrays.copyOfRange(inorder, 0, rootIdx);
        int[] inorderRight = Arrays.copyOfRange(inorder, rootIdx + 1, inorder.length);
        
        root.left = buildTree(preorderLeft, inorderLeft);
        root.right = buildTree(preorderRight, inorderRight);
        return root;
    }
}

/**
Approach 2: hashmap and pointers in array
TC: O(n)
SC: O(n) for extra space for hash map. Otherwise, it would have been height of tree.
*/
class Solution {
    int preOrderIndex = 0;
    HashMap<Integer, Integer> inorderIndexMap = null;
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        inorderIndexMap = new HashMap<>();
        
        for (int i = 0; i < inorder.length; ++i) {
            inorderIndexMap.put(inorder[i], i);
        }
        return arrayToTree(preorder, 0, preorder.length - 1);
    }
    
    private TreeNode arrayToTree(int[] preorder, int left, int right) {
        if (left > right) return null;
        
        TreeNode root = new TreeNode(preorder[preOrderIndex++]);
        root.left = arrayToTree(preorder, left, inorderIndexMap.get(root.val) - 1);
        root.right = arrayToTree(preorder, inorderIndexMap.get(root.val) + 1, right);
        return root;
    }
}